src/ray/common/scheduling/scheduling_alg.h

```diff
@@ -0,0 +1,147 @@
+  ///
+  /// \param local_node_id: ID of local node,
+  /// \param node_resources: The total and the available resources associated
+  /// with the local node.
+  ClusterResources(int64_t local_node_id, const NodeResources &node_resources);
+
+  /// Add a new node, or update the resources of an existing node.
+  ///
+  /// \param node_id: Node ID.
+  /// \param node_resources: Up to date total and available resources of the node.
+  void add(int64_t node_id, const NodeResources &node_resources);
+
+  /// Remove node from the cluster data structure. This happens
+  /// when a node fails or it is removed from the cluster.
+  ///
+  /// \param ID of the node to be removed.
+  bool remove(int64_t node_id);
+
+   /// Check whether a task request can be scheduled given a node's resources.
+   ///
+   ///  \param task_req: Task request to be scheduled.
+   ///  \param nr: Node's resources.
+   ///
+   ///  \return: -1, if the request cannot be scheduled. This happens when at
+   ///           least a hard constraints is violated.
+   ///           >= 0, the number soft constraint violations. If 0, no
+   ///           constraint is violatede.
+  int64_t isSchedulable(const TaskReq &task_req, const NodeResources &nr);
+
+   ///  Find a node in the cluster on which we can schedule a given task request.
+   ///
+   ///  First, this function checks whether the local node can schedule
+   ///  the request without violating any constraints. If yes, it returns the
+   ///  ID of the local node.
+   ///
+   ///  If not, this function checks whether there is another node in the cluster
+   ///  that satisfies all request's constraints (both soft and hard).
+   ///
+   ///  If no such node exists, the function checks whether there are nodes
+   ///  that satisfy all the request's hard constraints, but might violate some
+   ///  soft constraints. Among these nodes, it returns a node which violates
+   ///  the least number of soft constraints.
+   ///
+   ///  Finally, if no such node exists, return -1.
+   ///
+   ///  \param task_req: Task to be scheduled.
+   ///  \param violations: The number of soft constraint violations associated
+   ///                     with the node returned by this function (assuming
+   ///                     a node that can schedule task_req is found).
+   ///
+   ///  \return -1, if no node can schedule the current request; otherwise,
+   ///          return the ID of a node that can schedule the task request.
+  int64_t getSchedulableNode(const TaskReq &task_req, int64_t *violations);
+
+  /// Update the available resources of a node when a task request is
+  /// scheduled on the given node.
+  ///
+  /// \param node_id: ID of node on which request is being scheduled.
+  /// \param task_req: task request being scheduled.
+  ///
+  /// \return true, if task_req can be indeed scheduled on the node,
+  /// and false otherwise.
+  bool updateAvailableResources(int64_t node_id, const TaskReq &task_req);
```

Comment:
This name is a bit ambiguous - don't know if you're adding or subtracting resources. Maybe `SubtractTaskResources` or similar is better